File: java-samples/poslink/EposPosLinkExample/src/main/java/com/example/eposposlinkexample/teya/PrintUtils.java
package com.example.eposposlinkexample.teya;

import com.example.eposposlinkexample.models.ProductItem;
import com.example.eposposlinkexample.util.PriceUtils;

import com.teya.unifiedepossdk.models.Align;
import com.teya.unifiedepossdk.models.ReceiptRow;
import com.teya.unifiedepossdk.models.RowElement;
import com.teya.unifiedepossdk.models.TableHeaderCell;
import com.teya.unifiedepossdk.models.TableRow;
import com.teya.unifiedepossdk.models.Template;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

final class PrintUtils {

    private static final DateTimeFormatter TIMESTAMP = DateTimeFormatter.ofPattern("dd/MM/yy · HH:mm");

    private PrintUtils() {
    }

    static Template buildReceiptTemplate(List<ProductItem> items, int tipMinor) {
        return new Template(List.of(
                new ReceiptRow.Items(List.of(
                        new RowElement.Text("CUSTOMER RECEIPT", true, Align.Left),
                        new RowElement.Text(LocalDateTime.now().format(TIMESTAMP), true, Align.Right))),
                ReceiptRow.Spacer.INSTANCE,
                ReceiptRow.Divider.INSTANCE,
                itemsTable(items),
                ReceiptRow.Divider.INSTANCE,
                amountRow("TIP", tipMinor),
                amountRow("TOTAL", totalMinor(items) + tipMinor),
                new ReceiptRow.Item(new RowElement.QrCode("https://teya.com", Align.Center)),
                ReceiptRow.Spacer.INSTANCE,
                ReceiptRow.Spacer.INSTANCE,
                new ReceiptRow.Item(new RowElement.Text("Thank you", true, Align.Center))));
    }

    static Template buildBillTemplate(String title, List<ProductItem> items, int totalMinor) {
        return new Template(List.of(
                new ReceiptRow.Item(new RowElement.Text("BILL", true, Align.Center)),
                new ReceiptRow.Item(new RowElement.Text(title, false, Align.Center)),
                ReceiptRow.Spacer.INSTANCE,
                ReceiptRow.Divider.INSTANCE,
                itemsTable(items),
                ReceiptRow.Divider.INSTANCE,
                amountRow("TOTAL", totalMinor)));
    }

    private static ReceiptRow.Table itemsTable(List<ProductItem> items) {
        List<TableHeaderCell> header = List.of(
                new TableHeaderCell(new RowElement.Text("ITEM", true, Align.Left), 1f),
                new TableHeaderCell(new RowElement.Text("PRICE", true, Align.Right), 1f));

        List<TableRow> rows = items.stream()
                .filter(item -> item.getQuantity() > 0)
                .map(item -> new TableRow(List.of(
                        new RowElement.Text(
                                item.getQuantity() + "x " + item.product().name().toUpperCase(), false, Align.Left),
                        new RowElement.Text(PriceUtils.formatPrice(item.lineTotal()), false, Align.Right))))
                .toList();

        return new ReceiptRow.Table(header, rows);
    }

    private static ReceiptRow.Items amountRow(String label, int amountMinor) {
        return new ReceiptRow.Items(List.of(
                new RowElement.Text(label, true, Align.Left),
                new RowElement.Text(PriceUtils.formatMinor(amountMinor), true, Align.Right)));
    }

    private static int totalMinor(List<ProductItem> items) {
        return items.stream().mapToInt(item -> PriceUtils.toMinorUnits(item.lineTotal())).sum();
    }
}

File: android-samples/poslink/EposPosLinkExample/app/src/main/java/com/example/eposappexample/poslink/teya/PrintUtils.kt
package com.example.eposappexample.poslink.teya

import com.example.eposappexample.poslink.formatMinor
import com.example.eposappexample.poslink.formatPrice
import com.example.eposappexample.poslink.models.Product
import com.teya.unifiedepossdk.models.Align
import com.teya.unifiedepossdk.models.ReceiptRow
import com.teya.unifiedepossdk.models.RowElement
import com.teya.unifiedepossdk.models.TableHeaderCell
import com.teya.unifiedepossdk.models.TableRow
import com.teya.unifiedepossdk.models.Template
import com.teya.unifiedepossdk.poslink.models.tabs.TabSummary
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

object PrintUtils {
    fun buildCustomPrintTemplate(products: List<Product>, tip: Double) = Template(
        listOf(
            ReceiptRow.Items(
                items = listOf(
                    RowElement.Text(
                        text = "CUSTOMER RECEIPT",
                        align = Align.Left,
                        bold = true
                    ),
                    RowElement.Text(
                        text = SimpleDateFormat(
                            "dd/MM/yy · HH:mm",
                            Locale.getDefault()
                        ).format(Date()),
                        align = Align.Right,
                        bold = true
                    ),
                )
            ),

            ReceiptRow.Spacer,
            ReceiptRow.Divider,

            ReceiptRow.Table(
                headerCells = listOf(
                    TableHeaderCell(
                        element = RowElement.Text(
                            text = "ITEM",
                            bold = true,
                            align = Align.Left
                        ),
                        1f
                    ),
                    TableHeaderCell(
                        element = RowElement.Text(
                            text = "PRICE",
                            bold = true,
                            align = Align.Right
                        ),
                        1f
                    ),
                ),
                rows = products.map { product ->
                    TableRow(
                        cells = listOf(
                            RowElement.Text(
                                text = "${product.quantity}x ${product.name.uppercase()}",
                                align = Align.Left,
                                bold = true
                            ),
                            RowElement.Text(
                                text = formatPrice(product.price * product.quantity),
                                align = Align.Right,
                                bold = true
                            ),
                        )
                    )
                }
            ),

            ReceiptRow.Divider,

            ReceiptRow.Items(
                items = listOf(
                    RowElement.Text(
                        text = "TIP",
                        align = Align.Left,
                        bold = true
                    ),
                    RowElement.Text(
                        text = formatPrice(tip),
                        align = Align.Right,
                        bold = true
                    )
                )
            ),
            ReceiptRow.Items(
                items = listOf(
                    RowElement.Text(
                        text = "TOTAL",
                        align = Align.Left,
                        bold = true
                    ),
                    RowElement.Text(
                        text = formatPrice(products.sumOf { it.price * it.quantity } + tip),
                        align = Align.Right,
                        bold = true
                    )
                )
            ),

            ReceiptRow.Item(
                RowElement.QrCode(
                    url = "https://teya.com",
                    align = Align.Center
                )
            ),

            ReceiptRow.Spacer,
            ReceiptRow.Spacer,

            ReceiptRow.Item(
                RowElement.Text(
                    text = "Thank you",
                    align = Align.Center,
                    bold = true
                )
            ),
        )
    )

    fun buildTableBillTemplate(tab: TabSummary, items: List<Product>, totalMinor: Int) =
        Template(
            listOf(
                ReceiptRow.Item(
                    RowElement.Text(text = "BILL", align = Align.Center, bold = true)
                ),
                ReceiptRow.Item(
                    RowElement.Text(text = tab.tabName, align = Align.Center)
                ),
                ReceiptRow.Spacer,
                ReceiptRow.Divider,
                ReceiptRow.Table(
                    headerCells = listOf(
                        TableHeaderCell(
                            element = RowElement.Text(
                                text = "ITEM",
                                bold = true,
                                align = Align.Left
                            ),
                            1f
                        ),
                        TableHeaderCell(
                            element = RowElement.Text(
                                text = "PRICE",
                                bold = true,
                                align = Align.Right
                            ),
                            1f
                        ),
                    ),
                    rows = items.map { product ->
                        TableRow(
                            cells = listOf(
                                RowElement.Text(
                                    text = "${product.quantity}x ${product.name.uppercase()}",
                                    align = Align.Left
                                ),
                                RowElement.Text(
                                    text = formatPrice(product.price * product.quantity),
                                    align = Align.Right
                                ),
                            )
                        )
                    }
                ),
                ReceiptRow.Divider,
                ReceiptRow.Items(
                    items = listOf(
                        RowElement.Text(text = "TOTAL", align = Align.Left, bold = true),
                        RowElement.Text(
                            text = formatMinor(totalMinor),
                            align = Align.Right,
                            bold = true
                        ),
                    )
                ),
            )
        )
}